cpu/sam0_common/periph/freqm.c

```
@@ -0,0 +1,224 @@
+    }
+
+    return freqm_is_busy;
+}
+
+static void _freqm_nvic_init(void)
+{
+    /* Enable interrupt */
+    NVIC_EnableIRQ(FREQM_IRQn);
+}
+
+bool freqm_start(int blocking)
+{
+    bool status = false;
+
+    /* Check if measurement is already in progress */
+    if (!_freqm_is_busy()) {
+        /* Clear the Done Interrupt flag */
+        FREQM->INTFLAG.reg = (uint8_t)FREQM_INTFLAG_DONE_Msk;
+
+        /* Start measurement */
+        FREQM->CTRLB.reg = (uint8_t)FREQM_CTRLB_START_Msk;
+
+        status = true;
+    }
+
+    if (blocking) {
+        /* Block until measurement is done */
+        mutex_lock(&sync_lock);
+    }
+
+    return status;
+}
+
+int freqm_frequency_get(const struct freqm_config *config, uint32_t *result)
+{
+    uint64_t result_tmp = 0U;
+    int overflow_condition = 0;
+
+    /* Calculate measured frequency */
+    result_tmp = FREQM->VALUE.reg * (uint64_t)(sam0_gclk_freq(config->ref_clk_src));
+    result_tmp = result_tmp / config->ref_clk_cnt;
+
+    /* Return measured frequency */
+    *result = (uint32_t)result_tmp;
+
+    /* Read overflow status */
+    overflow_condition = ((int)FREQM->STATUS.reg & FREQM_STATUS_OVF_Msk);
+
+    /* Clear overflow status */
+    FREQM->STATUS.reg = (uint8_t)FREQM_STATUS_OVF_Msk;
+
+    return overflow_condition;
+}
+
+void freqm_callback_register(freqm_callback_t freqm_callback, void *context)
+{
+    freqm_obj.callback = freqm_callback;
+
+    freqm_obj.context = context;
+}
+
+void irq_freqm(void)
+{
+    void *context_var;
+
+    FREQM->INTFLAG.reg = (uint8_t)FREQM_INTFLAG_DONE_Msk;
+
+    if (freqm_obj.callback != NULL) {
+        if (_freqm_is_busy() == false) {
+            context_var = freqm_obj.context;
+            freqm_obj.callback(context_var);
+        }
+    }
+
+    mutex_unlock(&sync_lock);
+}
+
+int freqm_init(struct freqm_config const *config)
+{
+    uint8_t gclk_io_id;
+
+    /* Sanity check configuration */
+    if (config->ref_clk_src > GCLK_GEN_NUM_MSB) {
```

You can do 
```suggestion
    assert(config->ref_clk_src <= GCLK_GEN_NUM_MSB) {
```

because `config` is compile-time const.
There is no way to handle this error as it can only result from a wrong board config, so fail early and loud.